Add PlayerInsertLecternBookEvent [1.20 port]

NoahvdAa opened the pull request:
Ports #4245 (fixes #2286) to 1.18 and addresses comments. The diff
has been simplified a lot because the vanilla already passes the player
into the resolveBook and setBook methods.

Proximyst committed:
Add PlayerInsertLecternBookEvent

Machine-Maker commented on patches/server/0846-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,50 @@
+             WrittenBookItem.resolveBookComponents(book, this.createCommandSourceStack(player), player);
+         }
++        // Paper start
++        if (player != null && getLevel() != null) {
```
```suggestion
+        if (player != null && this.level != null && player instanceof ServerPlayer serverPlayer) {
```

Machine-Maker commented on patches/api/0354-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,73 @@
++    }
++
++    @NotNull
++    public Lectern getLectern() {
```
Probably a javadoc comment saying this is a not a snapshot blockstate, but a live one

Machine-Maker commented on patches/api/0354-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,73 @@
++    }
++
++    @NotNull
++    public ItemStack getBook() {
```
comment saying changes will be reflected on the itemstack

Machine-Maker commented on patches/server/0846-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,50 @@
++                org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(book)
++            );
++            if (!event.callEvent()) return ItemStack.EMPTY;
++            return ((org.bukkit.craftbukkit.inventory.CraftItemStack) event.getBook()).handle;
```
Since there is a setter for the book itemstack on the event, you cannot cast it to CraftItemStack since there is no guarantee that it will be an instance of it.

Machine-Maker commented on patches/server/0846-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,50 @@
++        if (player != null && getLevel() != null) {
++            io.papermc.paper.event.player.PlayerInsertLecternBookEvent event = new io.papermc.paper.event.player.PlayerInsertLecternBookEvent(
++                (org.bukkit.entity.Player) player.getBukkitEntity(),
++                (org.bukkit.block.Lectern) getLevel().getWorld().getBlockAt(worldPosition.getX(), worldPosition.getY(), worldPosition.getZ()).getState(false),
```
```suggestion
+                (org.bukkit.block.Lectern) CraftBlock.at(this.level, this.worldPosition).getState(false),
```

Machine-Maker commented on patches/server/0846-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,50 @@
++        // Paper start
++        if (player != null && getLevel() != null) {
++            io.papermc.paper.event.player.PlayerInsertLecternBookEvent event = new io.papermc.paper.event.player.PlayerInsertLecternBookEvent(
++                (org.bukkit.entity.Player) player.getBukkitEntity(),
```
```suggestion
+                serverPlayer.getBukkitEntity(),
```

Machine-Maker reviewed (changes requested):
I'm not sure the itemstack should be a mirror AND have a setter in the event. I think usually you use mirrors if you can't reassign the nms itemstack where the event is called or if the event doesn't have a setter, it seems odd to have both.

Machine-Maker commented on patches/server/0877-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,50 @@
+         if (!(Boolean) state.getValue(LecternBlock.HAS_BOOK)) {
+             if (!world.isClientSide) {
+                 LecternBlock.placeBook(player, world, pos, state, stack);
++                return ((LecternBlockEntity) world.getBlockEntity(pos)).hasBook(); // Paper
```
I would do an instance check on the result of world.getBlockEntity(pos) here as something else might cancel the block placement, or even in the PlayerInsertLecternBookEvent they change the block and remove the block entity. Like all over Mojang changes blocks, and then they check if the tile entity was created, they don't assume it was. Look at the first couple lines of placeBook

Machine-Maker commented on patches/api/0387-Add-PlayerInsertLecternBookEvent.patch:
```diff
@@ -0,0 +1,93 @@
++     *
++     * @param book the book to insert
++     */
++    public void setBook(@Nullable ItemStack book) {
```
I think we are trying to move away from nullable ItemStack in new stuff.

Owen1212055 reviewed (approved):
But see MM's comment

lynxplay committed:
Rebase

Additionally, moves the event call higher up and removes the lectern
block state clone from the event.

lynxplay committed:
jSpecify

lynxplay committed:
Shrink correctly

Machine-Maker committed:
style fixups

Machine-Maker committed:
move methods

Machine-Maker committed:
rename param to cancel

Machine-Maker committed:
fix javadocs

Machine-Maker committed:
more javadoc fixes

Machine-Maker committed:
fix co-author on patch from yesterday

Machine-Maker committed:
last fix
